dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java

```diff
@@ -2295,4 +2298,92 @@ public Set<OrcidAuthorInformation> getWorkflowVersionOrcidAuthors(@Parameter(hid
         return orcidAuthorInfo;
     }
 
+    /**
+     * Scans Zenodo for DOIs issued against GitHub repos with registered workflows in Dockstore, updating the Dockstore workflows
+     * with those DOIs.
+     *
+     * @param user
+     * @param filter - optional filter to scope to a single GitHub repository in the format myorg/myrepo
+     * @return
+     */
+    @POST
+    @RolesAllowed({"admin", "curator"})
+    @Path("/updateDois")
+    @UnitOfWork
+    @Timed
+    @Operation(operationId = "updateDois", description = "Searches Zenodo for DOIs referencing GitHub repos, and updates Dockstore entries with them", security = @SecurityRequirement(name = JWT_SECURITY_DEFINITION_NAME))
+    public List<Workflow> updateDois(@Parameter(hidden = true) @Auth User user,
+            @Parameter(description = "Optional GitHub full repository name, e.g., myorg/myrepo, to only update entries for that repository") @QueryParam("filter") String filter) {
+        final List<Workflow> publishedWorkflows = getPublishedGitHubWorkflows(filter);
+
+        // Get a map of GitHub repos to Dockstore workflows
+        final Map<String, List<Workflow>> repoToWorkflowsMap = publishedWorkflows.stream()
+                .filter(w -> w.getSourceControl() == SourceControl.GITHUB)
+                .collect(Collectors.groupingBy(w -> w.getOrganization() + '/' + w.getRepository()));
+
+        final List<ZenodoHelper.GitHubRepoDois> gitHubRepoDois = repoToWorkflowsMap.keySet().stream().sorted()
+                .map(ZenodoHelper::findDoisForGitHubRepo)
+                .flatMap(Collection::stream)
+                .toList();
+
+        final Set<Workflow> updatedWorkflows = new HashSet<>();
+        gitHubRepoDois.stream().forEach(gitHubRepoDoi -> {
+            final String repo = gitHubRepoDoi.repo();
+            final String conceptDoi = gitHubRepoDoi.conceptDoi();
+            final List<Workflow> workflows = repoToWorkflowsMap.get(repo);
+            workflows.stream().forEach(workflow -> {
+                final Doi existingGitHubDoi = workflow.getConceptDois().get(Doi.DoiInitiator.GITHUB);
+                if (existingGitHubDoi != null && conceptDoi.equals(existingGitHubDoi.getName())) {
+                    LOG.warn("Skipping DOI %s for workflow %s because it already has a Zenodo DOI from GitHub".formatted(conceptDoi, workflow.getWorkflowPath()));
+                } else {
+                    final boolean noDoiToStart = workflow.getConceptDois().isEmpty();
```

Comment:
We're on this side of the `if` because either there's no existing GitHub DOI, or the existing GitHub DOI doesn't match the one we found on Zenodo.  Can the latter case happen in practice, and if yes, should we be setting the GitHub DOI to the latest one we found?

Reply:
Yikes, I messed up my test on the previous line, it's supposed to be `!conceptDoi.equals(existingGitHubDoi.getName())`. I'll update and add a test for it.

I think it's possible, although unlikely. But if it happens, I think we should just leave what's there. 

It would mean there are 2 concept DOIs for the same repo issued by Zenodo-GitHub integration (or something like it). We have to chose one, and in case somebody already references the DOI, I think it's better to keep the existing one, although you could easily make an argument the other way around.